### PR TITLE
feat: shopify: ecom fix and enhancement

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

Resolve INT-360
> 
- Introducing a new Shopify version with strict ecomm spec according to rudderstack docs.
We will provide a toggle in the UI to use the new or old version.
With this old version would be deprecated.
- Added the event-specific mapping JSONs for ecomm events.

**Events changed (split/merged):**
- Split the **checkouts/update** webhook into **checkout_step_viewed**, **checkout_step_completed** and **payment_info_entered**

**Events Added:**
Product Added 
Product Removed

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request 

## Summary by CodeRabbit

- **New Features**
  - Search functionality now available with a new search bar for easier navigation and quicker access to information.
  - Enhanced data structure for Shopify integration, including new line items and webhook payload properties.
  - Additional utility functions for array operations, such as summing values and concatenating array elements.

- **Bug Fixes**
  - Resolved issues with Redis connection tests to ensure more reliable data fetching.
  - Corrected email extraction from payloads to improve data accuracy.

- **Documentation**
  - Updated test descriptions for clarity and better understanding of test purposes.

- **Refactor**
  - Improved Shopify integration with new utility functions for data handling and identifier resolution.
  - Streamlined event processing logic for Shopify events to enhance performance.

- **Tests**
  - Expanded test coverage for Shopify integration, including new scenarios for cart token extraction, identifier resolution, and event enrichment.

- **Chores**
  - General maintenance and codebase cleanup to ensure smooth operation and integration stability.

### CODE REVIEW

File: src/v1/sources/shopify/transform.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    // eslint-disable-next-line no-param-reassign
+    event = removeUndefinedAndNullValues(event);
+    return event;
+  });
+  return message;
+};
+
+const process = async (inputEvent) => {
+  const { event, source } = inputEvent;
+  if (source?.Config?.version !== 'v1') {
```
Comment (ItsSudip) [inline src/v1/sources/shopify/transform.js:71]:
```suggestion
  if (source.Config?.version !== 'v1') {
```

File: src/v1/sources/shopify/identifierEventsLayer.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
+const { IDENTIFIER_EVENTS, NO_OPERATION_SUCCESS } = require('./config');
+const stats = require('../../../util/stats');
+const { getLineItemsToStore } = require('./commonUtils');
+const { RedisDB } = require('../../../util/redis/redisConnector');
+const logger = require('../../../logger');
+
+const IdentifierEventLayer = {
+  isIdentifierEvent(event) {
+    return IDENTIFIER_EVENTS.includes(event?.event);
```
Comment (ItsSudip) [inline src/v1/sources/shopify/identifierEventsLayer.js:9]:
```suggestion
    return IDENTIFIER_EVENTS.includes(event.event);
```
Can we update the parameter name to message from event?

File: src/v1/sources/shopify/identityResolutionLayer.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+   * 3. This means we don't have `anonymousId` and hence events CAN NOT be stitched and we check for cartToken
+   *    a. if cartToken is available we return its hash value
+   *    b. else we check if the event is an SHOPIFY_ADMIN_ONLY_EVENT
+   *       -> if true we return `null`;
+   *       -> else we don't have any identifer (very edge case) we return `random anonymousId`
+   *    No Random SessionId is generated as its not a required field
+   * @param {*} message
+   * @param {*} metricMetadata
+   * @returns
+   */
```
Comment (ItsSudip) [inline src/v1/sources/shopify/identityResolutionLayer.js:33]:
```suggestion
   * This function retrieves anonymousId and sessionId in the following steps:
   * 1. Checks for `rudderAnonymousId`and `rudderSessionId in `note_atrributes`
   * 2. Checks in dbData
   * 3. This means we don't have `anonymousId` and hence events CAN NOT be stitched and we check for cartToken
   *    a. if cartToken is available we return its hash value
   *    b. else we check if the event is a SHOPIFY_ADMIN_ONLY_EVENT
   *       -> if true we return `null`;
   *       -> else we don't have any identifier (very edge case) we return `random anonymousId`
   *    No Random SessionId is generated as it's not a required field
   * @param {*} message
   * @param {*} metricMetadata
   * @returns
   */
```

File: test/apitests/data_scenarios/source/v1/Untitled-1.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+
```
Comment (utsabc) [inline test/apitests/data_scenarios/source/v1/Untitled-1.js:1]: File name not correct

File: test/apitests/data_scenarios/source/v1/Untitled-1.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        config.customActivityPrimaryKeyMap.forEach((item: { from: string; to: string }) => {
+            activityPrimaryKeyMap[item.from] = item.to;
+        });
+    }
+    if (config.customActivityEventMap) {
+        config.customActivityEventMap.forEach((item: { from: string; to: string }) => {
+            activityEventMap[item.from] = item.to;
+        });
+    }
+    Object.keys(activityEventMap).forEach(event => {
```
Comment (utsabc) [inline test/apitests/data_scenarios/source/v1/Untitled-1.js:33]: This is config migration script wrongly added

File: src/v1/sources/shopify/identityResolutionLayer.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        sessionId,
+      };
+    }
+    anonymousId = anonymousId || dbData?.anonymousId;
+    sessionId = sessionId || dbData?.sessionId;
+    if (!isDefinedAndNotNull(anonymousId)) {
+      /* anonymousId or sessionId not found from db as well
+            Hash the id and use it as anonymousId (limiting 256 -> 36 chars) and sessionId is not sent as its not required field
+            */
+      anonymousId = v5(cartToken, v5.URL);
```
Comment (ItsSudip) [inline src/v1/sources/shopify/identityResolutionLayer.js:64]: Can we add a stat here? It will help us find how many events the redisDB search has failed.

File: test/apitests/data_scenarios/source/v1/Untitled-1.js (hunk 5 above)
Comment (anantjain45823) [reply test/apitests/data_scenarios/source/v1/Untitled-1.js:33]: Yeah will remove it

File: src/v0/util/index.js [hunk 7]
```diff
@@ -614,6 +614,25 @@ const handleSourceKeysOperation = ({ message, operationObject }) => {
...
+    }
+    case 'concatenationInArray': {
+      const { propertyKey, arrayKey } = args[0];
+      const arrayValues = get(message, arrayKey);
+      if (lodash.isArray(arrayValues) && isNotEmpty(arrayValues)) {
+        result = arrayValues.map((item) => get(item, propertyKey, '')).join(', ');
+        return result;
+      }
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/util/index.js:635]: @Gauravudia can you take this

File: src/v1/sources/shopify/identityResolutionLayer.js (hunk 6 above)
Comment (anantjain45823) [reply src/v1/sources/shopify/identityResolutionLayer.js:64]: we already have that metric mentioned elsewhere

File: src/v1/sources/shopify/data/CheckoutStartedConfig.json [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      "type": "toNumber"
+    }
+  },
+  {
+    "sourceKeys": "subtotal_price",
+    "destKey": "revenue",
+    "metadata": {
+      "type": "toNumber"
+    }
+  },
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/data/CheckoutStartedConfig.json:22]: @Gauravudia can you check this ?

File: src/v1/sources/shopify/transform.js (hunk 1 above)
Comment (anantjain45823) [reply src/v1/sources/shopify/transform.js:71]: @ItsSudip as utsab suggested thats what it is about it will help us in redirecting the request without source object

File: src/v1/sources/shopify/transform.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    return processV0(event);
+  }
+  const metricMetadata = {
+    writeKey: event.query_parameters?.writeKey?.[0],
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/transform.js:75]: Okay then ideally we should be removing `writeKey` . So shall I replace `writeKey` with sourceId

File: src/v1/sources/shopify/commonUtils.test.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        error: 'Query_parameters is missing',
+      };
+      try {
+        getShopifyTopic(input);
+      } catch (error) {
+        expect(error.message).toEqual(expectedOutput.error);
+      }
+    });
+  });
+
```
Comment (utsabc) [inline src/v1/sources/shopify/commonUtils.test.js:60]: Can we add other methods of util to unit test cases

File: src/v1/sources/shopify/enrichmentLayer.test.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        },
+      };
+
+      const output = enrichPayload.setExtraNonEcomProperties(
+        message,
+        ordersCancelledWebhook,
+        'orders_cancelled',
+      );
+      expect(output).toEqual(expectedOutput);
+    });
```
Comment (utsabc) [inline src/v1/sources/shopify/enrichmentLayer.test.js:53]: Can we add more edge case scenarios if possible (use codium)

File: src/v1/sources/shopify/identifierEventsLayer.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
+const { IDENTIFIER_EVENTS, NO_OPERATION_SUCCESS } = require('./config');
+const stats = require('../../../util/stats');
+const { getLineItemsToStore } = require('./commonUtils');
+const { RedisDB } = require('../../../util/redis/redisConnector');
+const logger = require('../../../logger');
+
+const IdentifierEventLayer = {
```
Comment (utsabc) [inline src/v1/sources/shopify/identifierEventsLayer.js:7]: Please add unit tests for the functions of this utility layer

File: src/v1/sources/shopify/identifyEventsLayer.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
+const Message = require('../../../v0/sources/message');
+const { EventType } = require('../../../constants');
+const {
+    INTEGRATION,
+    MAPPING_CATEGORIES
+} = require('./config');
+
+const identifyLayer = {
```
Comment (utsabc) [inline src/v1/sources/shopify/identifyEventsLayer.js:8]: Please add unit tests for this function

File: src/v1/sources/shopify/transform.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (utsabc) [reply src/v1/sources/shopify/transform.js:75]: No just add it

File: src/v1/sources/shopify/data/CheckoutStartedConfig.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v1/sources/shopify/data/CheckoutStartedConfig.json:22]: @ItsSudip The [link](https://shopify.dev/docs/api/liquid/objects/order#order-subtotal_line_items) you have provided is for liquid objects.

File: src/v1/sources/shopify/transform.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/transform.js:75]: okay

File: src/v1/sources/shopify/transform.js [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    event = removeUndefinedAndNullValues(event);
+    return event;
+  });
+  return message;
+};
+
+const process = async (inputEvent) => {
+  const { event, source } = inputEvent;
+  if (source?.Config?.version !== 'v1') {
+    return processV0(event);
```
Comment (utsabc) [inline src/v1/sources/shopify/transform.js:72]:
```suggestion
    return await processV0(event);
```

File: src/v0/util/index.js [hunk 15]
```diff
@@ -614,6 +614,25 @@ const handleSourceKeysOperation = ({ message, operationObject }) => {
         }
       }
       return result;
+    case 'additionInArray': {
+      result = 0;
+      const { propertyKey, arrayKey } = args[0];
+      const arrayValues = get(message, arrayKey);
+      if (lodash.isArray(arrayValues) && isNotEmpty(arrayValues)) {
+        result = arrayValues.reduce((acc, item) => acc + get(item, propertyKey, 0), 0);
+        return result;
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v0/util/index.js:623]: @Gauravudia can you check this?

[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/util/index.js:631]: Added a check for isNaN

File: src/v1/sources/shopify/data/CheckoutStartedConfig.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v1/sources/shopify/data/CheckoutStartedConfig.json:22]: **subtotal_price** is a price in [Shopify](https://shopify.dev/docs/api/admin-rest/2023-10/resources/checkout#:~:text=won%27t%20be%20displayed.-,subtotal_price,-read%2Donly) doc 
`The price of the checkout in presentment currency before shipping, taxes, and tips.`
and we are mapping it with **revenue** which is `Revenue associated with the transaction, excluding the shipping and tax details`
@utsabc should we remove this mapping from transformer?

File: src/v0/util/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/util/index.js]: added the testcases

**Comment by anantjain45823** [comment]
@coderabbitai review

File: test/apitests/data_scenarios/source/v1/Untitled-1.js (hunk 4 above)
Comment (anantjain45823) [reply test/apitests/data_scenarios/source/v1/Untitled-1.js]: removed the file

File: src/v1/sources/shopify/transform.js [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  }
+  const metricMetadata = {
+    writeKey: event.query_parameters?.writeKey?.[0],
+    sourceId: source.ID,
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/transform.js:76]: We will need to change it for all metrics then I will change it

File: src/v1/sources/shopify/data/identifyMapping.json [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+    "sourceKeys": "verified_email",
+    "destKeys": "traits.verifiedEmail"
+  },
+  {
+    "sourceKeys": "multipass_identifier",
+    "destKeys": "traits.multipassIdentifier"
+  },
+  {
+    "sourceKeys": "tax_exempt",
+    "destKeys": "traits.taxExempt"
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/data/identifyMapping.json]: Done

File: src/v1/sources/shopify/transform.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/transform.js:76]: Done

File: src/v1/sources/shopify/enrichmentLayer.test.js (hunk 11 above)
Comment (anantjain45823) [reply src/v1/sources/shopify/enrichmentLayer.test.js:53]: Done

File: src/v1/sources/shopify/trackEventsLayer.js [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+      Object.keys(prevLineItems).forEach((lineItemID) => {
+        const product = prevLineItems[lineItemID];
+        const updatedProduct = this.getUpdatedProductProperties(product, event?.id || event?.token);
+        updatedProduct.id = lineItemID;
+        events.push(this.ecomPayloadBuilder(updatedProduct, 'product_removed'));
+      });
+    }
+    if (events.length > 0) {
+      await this.updateCartState(
+        getLineItemsToStore(event),
```
[the remark this answers is not in the document]
Comment (anantjain45823) [reply src/v1/sources/shopify/trackEventsLayer.js]: Changed the param name

**Comment by devops-github-rudderstack** [comment]
This PR is considered to be stale. It has been open for 20 days with no further activity thus it is going to be closed in 7 days. To avoid such a case please consider removing the stale label manually or add a comment to the PR.
